source/pele/pspin_spherical.h

```diff
@@ -0,0 +1,212 @@
+#ifndef _P_SPIN_SPHERICAL_H_
+#define _P_SPIN_SPHERICAL_H_
+
+#include "array.h"
+#include "base_potential.h"
+
+#include <iostream>
+#include <cassert>
+#include <cmath>
+#include "pele/meta_pow.h"
+#include "pele/combination.h"
+
+namespace pele {
+
+template <size_t p>
+class MeanFieldPSpinSpherical : public BasePotential {
+protected:
+    pele::Array<double> m_interactions, m_spins;
+    pele::Array<size_t> m_indexes;
+    size_t m_N;
+    double m_tol;
+    static const size_t m_p = p;
+    pele::combination_generator<size_t*> m_combination_generator;
+    pele::combination_generator<size_t*> m_combination_generator_grad;
+    pele::combination_generator<size_t*> m_combination_generator_hess;
+    size_t m_get_index(size_t* comb){
+        size_t idx = comb[0];
+        size_t d = m_N;
+        for (size_t i = 1; i < m_p; ++i){
+            idx += comb[i] * d;
+            d *= m_N;
+        }
+        return idx;
+    }
+
+    //this function normalizes the spin vector and orthogonalises the gradient to it
+    inline void m_orthonormalize(Array<double>& spins_, Array<double>& grad)
+    {
+        if (spins_.size() != grad.size()) {
+            throw std::invalid_argument("grad.size() be the same as spin_.size()");
+        }
+
+        bool success = true;
+        Array<double> spins = spins_.copy();
+
+        grad /= norm(grad);
```

Comment:
yes you're right I fixed this lat night but haven't pushed it.

Now the problem that I am seeing when actually performing a minimization is
that if I don't normalize the spins (sum s_i^2 = N) at each step I get the
wrong energies and minimization never converges because the rms settles to
some value much greater than the usual tolerance. However, when I normalize
the spin at each energy call I find the global minimum but while lbfgs' rms
converges fire's rms still doesn't. Anyway the function is complicated now,
maybe there is another mistake and replacing with a single line as you
suggest might get rid of the problem.

On Tue, Dec 1, 2015 at 8:15 AM, Jacob Stevenson <notifications@github.com>
wrote:

> In source/pele/pspin_spherical.h
> <https://github.com/pele-python/pele/pull/138#discussion_r46249877>:
>
> > +            d *= m_N;
> > +        }
> > +        return idx;
> > +    }
> > +
> > +    //this function normalizes the spin vector and orthogonalises the gradient to it
> > +    inline void m_orthonormalize(Array<double>& spins_, Array<double>& grad)
> > +    {
> > +        if (spins_.size() != grad.size()) {
> > +            throw std::invalid_argument("grad.size() be the same as spin_.size()");
> > +        }
> > +
> > +        bool success = true;
> > +        Array<double> spins = spins_.copy();
> > +
> > +        grad /= norm(grad);
>
> why are you normalizing the gradient? this doesn't seem correct. You want
> to subtract the component parallel to the spin vector. Can't this whole
> function be replaced by the below line?
>
> grad -= dot(grad, spins) / norm(spins)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pele-python/pele/pull/138/files#r46249877>.
>